union_pooling/union_pooling/union_pooler.py

```diff
@@ -0,0 +1,247 @@
+                                      stimulusThreshold,
+                                      synPermInactiveDec,
+                                      synPermActiveInc,
+                                      synPermConnected,
+                                      minPctOverlapDutyCycle,
+                                      minPctActiveDutyCycle,
+                                      dutyCyclePeriod,
+                                      maxBoost,
+                                      seed,
+                                      spVerbosity,
+                                      wrapAround)
+
+    self._activeOverlapWeight = activeOverlapWeight
+    self._predictedActiveOverlapWeight = predictedActiveOverlapWeight
+    self._maxUnionActivity = maxUnionActivity
+    self._synPredictedInc = synPredictedInc
+
+    self._poolingActivation = numpy.zeros(self._numColumns, dtype="int32")
+    self._unionSDR = []
+
+
+  def reset(self):
+    """
+    Reset the state of the Union Pooler.
+    """
+
+    # Reset Union Pooler fields
+    self._poolingActivation = numpy.zeros(self._numColumns, dtype="int32")
+    self._unionSDR = []
+
+    # Reset Spatial Pooler fields
+    self._overlapDutyCycles = numpy.zeros(self._numColumns, dtype=REAL_DTYPE)
+    self._activeDutyCycles = numpy.zeros(self._numColumns, dtype=REAL_DTYPE)
+    self._minOverlapDutyCycles = numpy.zeros(self._numColumns, dtype=REAL_DTYPE)
+    self._minActiveDutyCycles = numpy.zeros(self._numColumns, dtype=REAL_DTYPE)
+    self._boostFactors = numpy.ones(self._numColumns, dtype=REAL_DTYPE)
+
+
+  def compute(self, activeInput, predictedActiveInput, burstingColumns,
+              learn):
+    """
+    Computes one cycle of the Union Pooler algorithm.
+    """
+    assert (numpy.size(activeInput) == self._numInputs)
+    assert (numpy.size(predictedActiveInput) == self._numInputs)
+    self._updateBookeepingVars(learn)
+
+    activeOverlaps = self._calculateOverlap(activeInput)
+    predictedActiveOverlaps = self._calculateOverlap(predictedActiveInput)
+    totalOverlap = (activeOverlaps * self._activeOverlapWeight  +
+                    predictedActiveOverlaps *
+                    self._predictedActiveOverlapWeight)
+
+    # Apply boosting when learning is on
+    if learn:
+      boostedOverlaps = self._boostFactors * totalOverlap
+    else:
+      boostedOverlaps = totalOverlap
+
+    # Apply inhibition to determine the winning cells
+    activeCells = self._inhibitColumns(boostedOverlaps)
+
+    if learn:
+      self._adaptSynapses(activeInput, predictedActiveInput, activeCells)
+      self._updateDutyCycles(totalOverlap, activeCells)
+      self._bumpUpWeakColumns()
+      self._updateBoostFactors()
+      if self._isUpdateRound():
+        self._updateInhibitionRadius()
+        self._updateMinDutyCycles()
+
+    # Update and return the Union SDR
+
+    # Decrement pooling activation of all cells
+    self._poolingActivation -= 1
```

Comment:
So would this version build up density until it reaches a max, and then stay there?